### ISSUE TITLE
(part 2) When a package references a tarball, and the tarball's dependencies change, the change isn't reflected in the shrinkwrap

### ISSUE THREAD

**Issue by pgonzal** [pnpm/pnpm#1362]
### Code to reproduce the issue:

The end user experience is the same as #1342, however I am opening a separate issue ("part 2") because 
`--no-prefer-frozen-shrinkwrap` flag did fix the previous repro, and the details of this repro are slightly different.

So far I don't have an isolated repro yet, but the overall steps go like this:

1. (I deleted **pnpmfile.js** to avoid any external influences.)  We use Rush to generate a file **projects/sp-application-base.tgz** with a **package.json** like this:

```json
{
  "name": "@rush-temp/sp-application-base",
  "version": "0.0.0",
  "private": true,
  "dependencies": {
    "@ms/odsp-utilities-bundle": "4.2.8",
    "@uifabric/styling": "5.33.0",
    "@types/chai": "3.4.34",
    "@types/mocha": "2.2.38",
    "chai": "~3.5.0",
    "gulp": "~3.9.1",
    "sinon": "~1.17.6",
    "@types/sinon": "1.16.34",
    "ts-jest": "~22.4.3",
    "@types/es6-promise": "0.0.33",
    "@types/webpack-env": "1.13.1",
    "@microsoft/load-themed-styles": "1.8.7",
    "tslib": "~1.8.0"
  }
}
```

Recall that this is referenced by **common/temp/package.json** like this:

```
{
  "dependencies": {
    "@rush-temp/sp-application-base": "file:./projects/sp-application-base.tgz",
    . . .
  },
  "description": "Temporary file generated by the Rush tool",
  "name": "rush-common",
  "private": true,
  "version": "0.0.0",
  . . .
}
```

2. Next, Rush runs this command:

```
pnpm install --no-optional --store pnpm-store --no-lock --no-prefer-frozen-shrinkwrap
```

...which creates a **shrinkwrap.yaml** file like this:

```yaml
dependencies:
  . . .
  '@rush-temp/sp-application-base': 'file:projects/sp-application-base.tgz'
  . . .
  'file:projects/sp-application-base.tgz':
    dependencies:
      '@microsoft/load-themed-styles': 1.8.7
      '@ms/odsp-utilities-bundle': 4.2.8
      '@types/chai': 3.4.34
      '@types/es6-promise': 0.0.33
      '@types/mocha': 2.2.38
      '@types/sinon': 1.16.34
      '@types/webpack-env': 1.13.1
      '@uifabric/styling': 5.33.0
      chai: 3.5.0
      gulp: 3.9.1
      sinon: 1.17.7
      ts-jest: 22.4.6
      tslib: 1.8.1
    dev: false
    name: '@rush-temp/sp-application-base'
    resolution:
      integrity: sha512-hO6OFJ+pJjIfC/vrofHr492vNIYeJwmHtHDBfgenmDRvvcyKvJd51HtAh8dFabcndGa7AWunYIkUQ/tw4oh1eA==
      tarball: 'file:projects/sp-application-base.tgz'
    version: 0.0.0
  . . .
```

3. Then we update the **package.json** in **sp-application-base.tgz** to have a newer version 5.33.1 of **@uifabric/styling**:

```json
{
  "name": "@rush-temp/sp-application-base",
  "version": "0.0.0",
  "private": true,
  "dependencies": {
    "@ms/odsp-utilities-bundle": "4.2.8",
    "@uifabric/styling": "5.33.1",
    "@types/chai": "3.4.34",
    "@types/mocha": "2.2.38",
    "chai": "~3.5.0",
    "gulp": "~3.9.1",
    "sinon": "~1.17.6",
    "@types/sinon": "1.16.34",
    "ts-jest": "~22.4.3",
    "@types/es6-promise": "0.0.33",
    "@types/webpack-env": "1.13.1",
    "@microsoft/load-themed-styles": "1.8.7",
    "tslib": "~1.8.0"
  }
}
```

4. Then we delete **shrinkwrap.yaml**.  We also delete the entire **node_modules** folder.

5. We run the same PNPM command line from above, which generates a new **shrinkwrap.yaml** file.  The integrity hash has changed (possibly due to tarball timestamps?), but the **@uifabric/styling** version is still 5.33.0.

```yaml
'file:projects/sp-application-base.tgz':
    dependencies:
      '@microsoft/load-themed-styles': 1.8.7
      '@ms/odsp-utilities-bundle': 4.2.8
      '@types/chai': 3.4.34
      '@types/es6-promise': 0.0.33
      '@types/mocha': 2.2.38
      '@types/sinon': 1.16.34
      '@types/webpack-env': 1.13.1
      '@uifabric/styling': 5.33.0
      chai: 3.5.0
      gulp: 3.9.1
      sinon: 1.17.7
      ts-jest: 22.4.6
      tslib: 1.8.1
    dev: false
    name: '@rush-temp/sp-application-base'
    resolution:
      integrity: sha512-hO6OFJ+pJjIfC/vrofHr492vNIYeJwmHtHDBfgenmDRvvcyKvJd51HtAh8dFabcndGa7AWunYIkUQ/tw4oh1eA==
      tarball: 'file:projects/sp-application-base.tgz'
    version: 0.0.0
```

#### Expected behavior: **@uifabric/styling** should be 5.33.1.

6. In this state, if I delete **shrinkwrap.yaml** and run the command again, the version will still be 5.33.0.

7. The problem seems to be with the cached entry **pnpm-store/2/local/D%3A%2FGitRepos%2Fspc%2Fcommon%2Ftemp%2Fprojects%2Fsp-application-base.tgz/package.json**, which still has the old contents:

```json
{
  "name": "@rush-temp/sp-application-base",
  "version": "0.0.0",
  "private": true,
  "dependencies": {
    "@ms/odsp-utilities-bundle": "4.2.8",
    "@uifabric/styling": "5.33.0",
    "@types/chai": "3.4.34",
    "@types/mocha": "2.2.38",
    "chai": "~3.5.0",
    "gulp": "~3.9.1",
    "sinon": "~1.17.6",
    "@types/sinon": "1.16.34",
    "ts-jest": "~22.4.3",
    "@types/es6-promise": "0.0.33",
    "@types/webpack-env": "1.13.1",
    "@microsoft/load-themed-styles": "1.8.7",
    "tslib": "~1.8.0"
  }
}
```

**Workaround**:  Delete the folder **pnpm-store/2/local** and delete **shrinkwrap.yaml** again. Then run the install command, and this time the **shrinkwrap.yaml** is updated with the correct version.

### Additional information:

 - `node -v` prints: 8.9.4
 - Windows, OS X, or Linux?: Windows

**Comment by pgonzal** [comment]
Here is an isolated repro:
[pnpm-bug.zip](https://github.com/pnpm/pnpm/files/2358702/pnpm-bug.zip)

1. Unzip the **pnpm-bug** folder.
2. Copy **projects/sp-application-base-step1.tgz** to be **projects/sp-application-base.tgz**
3. Run `pnpm install --store pnpm-store --no-prefer-frozen-shrinkwrap`
4.  Copy **projects/sp-application-base-step2.tgz** to be **projects/sp-application-base.tgz**
5. Delete **shrinkwrap.yaml**
6. Run `pnpm install --store pnpm-store --no-prefer-frozen-shrinkwrap`

Expected: **shrinkwrap.yaml** references **@uifabric/styling** version 5.33.1.

Actual: **shrinkwrap.yaml** references **@uifabric/styling** version 5.33.0.

```yaml
  'file:projects/sp-application-base.tgz':
    dependencies:
      '@uifabric/styling': 5.33.0
    dev: false
    name: '@rush-temp/sp-application-base'
    resolution:
      integrity: sha512-JUmAX2t1eTBU8PkyYNMBE+ltP45sVYHLh8frHp+H8OV7UdBq8r+uokVBhG/NBW7NIoMTGrAQCPZFoTICYU87cA==
      tarball: 'file:projects/sp-application-base.tgz'
    version: 0.0.0
```

**Comment by pgonzal** [comment]
This issue does not repro with PNPM 2.6.2 (but it does repro if steps 1-3 were performed using PNPM 2.15.0).  Apparently it is a recent regression.

**Comment by zkochan** [comment]
This seems like not a regression. It was working this way from the beginning. But it should be easy to fix.

When there is a `shrinkwrap.yaml` file, pnpm compares the integrity of the tarball to the integrity that was saved in the lockfile. If they don't match, pnpm reunpacks the tarball to the store.

However, when there is no `shrinkwrap.yaml` file (in your steps you remove it), pnpm only checks if there is an unpacked version of the tarball in the store. pnpm is not checking whether the unpacked files were from a tarball with the same integrity as the current tarball. (Also, the path to the unpacked tarball is the encoded path of the tarball)

I think the right solution would be to use the tarball's integrity as [package id](https://github.com/pnpm/spec/blob/master/shrinkwrap/3.9.md#packagesdependencypathid) of local tarball dependencies. That way, if the tarball's integrity changes, the location of the unpacked tarball changes in the store

**Comment by pgonzal** [comment]
> This seems like not a regression. It was working this way from the beginning.

Why does the issue NOT repro when using PNPM version 2.6.2?

**Comment by zkochan** [comment]
:ship: [2.15.1](https://github.com/pnpm/pnpm/releases/tag/v2.15.1)

**Comment by pgonzal** [comment]
(Fixed in PR https://github.com/pnpm/pnpm/pull/1363)

**Comment by pgonzal** [comment]
Thank you very much!  I have confirmed that this issue is fixed in 2.15.1.
